demos/bounded_heat/bounded_heat.py.rst

```diff
@@ -0,0 +1,202 @@
+
+    v_c = TestFunction(V)
+
+    F_c = (inner(Dt(u_c), v_c) * dx + inner(grad(u_c), grad(v_c)) * dx - inner(rhs, v_c) * dx)
+
+We use exact boundary conditions in both cases: ::
+
+    bc = DirichletBC(V, uexact, "on_boundary")
+
+For the unconstrained approximation, we configure the :class:`.TimeStepper` in a 
+familiar way: ::
+
+    stepper = TimeStepper(F, butcher_tableau, t, dt, u, bcs=bc, solver_parameters=lu_params)
+
+We will enforce nonnegativity when finding the constrained approximation. We now set up the keyword database to 
+configure an instance of :class:`.TimeStepper` for this task. We first specify, using the 
+keyword ``stage_type``, that we wish to use a stage-value formulation of the underlying collocation 
+method. The keyword ``basis_type`` then allows us to change the basis of the collocation 
+polynomial to the Bernstein basis. Having done this, we must specify a solver which is able to handle bounds 
+constraints. In this example we solve a variational inequality using ``vinewtonrsls`` by passing ``vi_params`` 
+as ``solver_parameters`` to the :class:`.TimeStepper`: ::
+
+    kwargs_c = {"stage_type": "value",
+                "basis_type": 'Bernstein',
+                "solver_parameters": vi_params
+            }
+
+    stepper_c = TimeStepper(F_c, butcher_tableau, t, dt, u_c, bcs=bc, **kwargs_c)
+
+We set the bounds as follows: ::
+
+    lb = Function(V)
+    lb.assign(0)
+
+    ub = None
+
+    bounds = ('stage', lb, ub)
+
+Passing the :meth:`~.TimeStepper.advance` method ``bounds`` will enforce the 
+given bounds on the coefficients of the collocation polynomial (now represented in the Bernstein basis). This 
+enforces the lower bound of :math:`0` uniformly in time.
+
+We now advance both semidiscrete systems in the usual way. We add the bounds as an argument 
+to the :meth:`~.TimeStepper.advance` method for the constrained approximation.  
+
+In order to monitor our approximate solutions, we check the minimum value of each after every step in time. 
+If an approximate solution violates the lower bound, we append a tuple to indicate the time and minimum value. ::
+
+    mins = []
+    mins_c = []
+
+    while (float(t) < float(Tf)):
+
+        if (float(t) + float(dt) > float(Tf)):
+            dt.assign(float(Tf) - float(t))
+
+        stepper.advance()
+        stepper_c.advance(bounds=bounds)
+
+        t.assign(float(t) + float(dt))
+
+        minv = min(u.dat.data)
+        if minv < 0:
+            mins.append((float(t), minv))
+
+        minv_c = min(u_c.dat.data)
+        if minv_c < 0:
+            mins_c.append((float(t), minv_c))
+
+        print(float(t))
+  
+Finally, we print the relative :math:`L^2` error and the time and severity (if any) of constraint violations: ::
+
+    np.set_printoptions(legacy='1.25')
+
+    print()
+    print(norm(u - uexact) / norm(uexact))
+    print(norm(u_c - uexact) / norm(uexact))
+    print()
+    print(mins)
```

Comment:
Also, the output is probably more descriptive/readable if you print out the discrete times as well, and only print the time levels where the bounds are violated.